ref: Remove distiction between free vector and free track parameters

niermann999 opened the pull request:
The free vector and the free track parameters are pretty much the same now, so let's use the classes public interface to access the data. This leads to some code simplifications in the track helper and the single parameter access

Also fixed a bug, where the CUDA code would not compile due to missing initializers on the mask values.

beomki-yeo commented:
I don't think I like this. I would keep the symmetry in the interface for bound and free track parameters

niermann999 commented:
> I don't think I like this. I would keep the symmetry in the interface for bound and free track parameters

I also removed it for bound track parameters, I just split the PR

beomki-yeo commented:
So for example, the next PR will have a function for `free_to_bound_param`? I am afraid that it is not possible to obtain the bound covariance from the free track paramaters which does not have covariance anymore

niermann999 commented:
> So for example, the next PR will have a function for `free_to_bound_param`? I am afraid that it is not possible to obtain the bound covariance from the free track paramaters which does not have covariance anymore

If it works better for you, I can absolutely rename ```bound_to_free_param``` again to ```bound_to_free_vector``` to make the name symmetric. In this case, we would return the raw free vector and just move it into free track parameters in the constructor, I suppose:
```
free_track_parameters<algebra_t> free_param{detail::bound_to_free_vector<frame_t>(trf, mask, bound_params)};
```
But, I was hoping I would not need to deal with this extra step anymore?

The next PR was meant to just wrap the 6-dim bound raw vector behind an interface as well, so that the track helper is no longer needed to interpret the data passed to clients. This was simply meant to improve "Information Hiding" in our code and abstract the internal implementation of the track parameters better from its clients. This is generally considered a helpful thing, because it increases maintainability by confining possible changes of the implementation of a class to the code of the class itself. 
For example: the track helper is not actually always used everywhere consistently (e.g. in the tests), because the raw vector is too exposed and direct access with the element getter is an obvious thing to do. This means that the way the data is accessed is not really centralized in a single place right now and we would need to touch a lot of code if anything changes there. And on future code, people might not be disciplined enough to always use the track helper.
Furthermore, information like ```p_T``` can now be accessed on the track parameter vector directly, without the need for any additional helper at all.

I did indeed have a ```free_to_bound_param``` function initially, where the covariance needed to be provided by the caller (as you explained above, this is expected) and e.g. the tracking surface could insert its barcode/transform automatically (```sf.free_to_bound_param(gctx, free_param, cov)```). But we hardly use the tracking functionality of the surface anyway and I was not sure that the following added much to the table:
```
bound_track_parameters<algebra_t> bound_param = detail::free_to_bound_param<frame_t>(trf, barcode, free_param, cov);
```
vs just
```
bound_track_parameters<algebra_t> bound_param{barcode, detail::free_to_bound_vector<frame_t>(trf, free_param), cov};
```
It's about equally verbose, so I removed it again.

beomki-yeo commented:
`track_helper` can be removed as I also don't see a great need to keep it

The problem of `bound_track_parameters<algebra_t> bound_param = detail::free_to_bound_param<frame_t>(trf, barcode, free_param, cov);` is that we don't calculate the free covariance explicitely anymore. So it is already difficult to keep the symmetrical interface.

niermann999 commented:
> `track_helper` can be removed as I also don't see a great need to keep it
> 
> The problem of `bound_track_parameters<algebra_t> bound_param = detail::free_to_bound_param<frame_t>(trf, barcode, free_param, cov);` is that we don't calculate the free covariance explicitely anymore. So it is already difficult to keep the symmetrical interface.

OK, so, do you want me apply my first suggestion in from above then, or what does this commen mean?

" I can absolutely rename ```bound_to_free_param``` again to ```bound_to_free_vector``` to make the name symmetric. In this case, we would return the raw free vector and just move it into free track parameters in the constructor, I suppose:
```
free_track_parameters<algebra_t> free_param{detail::bound_to_free_vector<frame_t>(trf, mask, bound_params)};
```
"
As, I also said in that comment, I removed ```detail::free_to_bound_param``` anyway, so it is not even a part of #794 anymore. Or was your argument that I should add it back?

beomki-yeo commented:
What I mean is that naming to `free_to_bound_vector` is better than `free_to_bound_param` even though that will make the command more verbose

beomki-yeo commented on core/include/detray/tracks/free_track_parameters.hpp:
```diff
@@ -28,47 +27,42 @@ struct free_track_parameters {
-    free_track_parameters()
-        : m_vector(matrix_operator().template zero<e_free_size, 1>()){};
+    /// Default constructor
+    free_track_parameters() = default;
```
or we can name this to `free_vector()` and never use free_track_parameter. That will harmonizer better with the next PR

niermann999 replied:
Yes, absolutely

niermann999 replied:
I renamed the class/transform function (```free_track_parameters```->```free_parameters_vector``` and ```bound_to_free_param```->```bound_to_free_vector```), but I was not so sure about completely removing the name ```free_track_parameters``` as the symmetric type to ```bound_track_parameters```. So I added an alias, which makes clear, what free track parameters are:
```
/// The free track parameters consist only of the parameter vector
template <typename algebra_t>
using free_track_parameters = free_parameters_vector<algebra_t>;
```
So, now it is completely symmetric in its implementation to the bound track parameters and just not adding the covariance etc. It will also cause less changes in downstream code.

Does that work for you?
